src/core/SampleBuffer.cpp

```diff
@@ -131,6 +131,79 @@ SampleBuffer::SampleBuffer( const f_cnt_t _frames )
 
 
 
+SampleBuffer::SampleBuffer(SampleBuffer& orig):
+	m_audioFile(orig.m_audioFile),
+	m_origData(MM_ALLOC(sampleFrame, orig.m_origFrames)),
+	m_origFrames(orig.m_origFrames),
+	m_data(MM_ALLOC(sampleFrame, orig.m_frames)),
+	m_frames(orig.m_frames),
+	m_startFrame(orig.m_startFrame),
+	m_endFrame(orig.m_endFrame),
+	m_loopStartFrame(orig.m_loopStartFrame),
+	m_loopEndFrame(orig.m_loopEndFrame),
+	m_amplification(orig.m_amplification),
+	m_reversed(orig.m_reversed),
+	m_frequency(orig.m_frequency),
+	m_sampleRate(orig.m_sampleRate)
+{
+	orig.m_varLock.lockForRead();
+	//Deep copy m_origData and m_data from original
+	memcpy(m_origData, orig.m_origData, m_origFrames* BYTES_PER_FRAME);
+	memcpy(m_data, orig.m_data, m_frames* BYTES_PER_FRAME);
+	orig.m_varLock.unlock();
+}
+
+
+
+
+void swap(SampleBuffer& first, SampleBuffer& second) noexcept
+{
+	using std::swap;
+
+	// Lock both buffers for writing
+	// TODO: No guaranteed ordering, risk of deadlock?
```

Comment:
I'm not sure I see the possibility of a deadlock being caused by the order those are called (them being different objects). But I do see a possibility of a deadlock being caused by the swap method being called from inside another method that already locked the object for writing, since right now the lock is not recursive. I see two options to handle this:

1) Adding an explicit initialization of the QReadWriteLock making it recursive on the SampleBuffer constructor. That way if we call the copy assignment (or swap method) from a method that locked for write, we wouldn't deadlock. We also would have to be careful to not call the copy assignment from inside a method that locked for reading without unlocking it first (because the recursive lock only works for locks of the same type: write->write->write or read->read->read, __not__ read->write).

2) Keeping the lock non-recursive and moving the responsibility of unlocking before calling the copy assignment to the method that calls it.

I think @DomClark and @PhysSong might have better insight than me on that one though.

Reply:
Thread A calls `swap(x, y)` while Thread B calls `swap (y, x)`, each thread grabs one lock and waits for the other. I thought I remembered these locks already being recursive, but you're right that I should make sure they are if they aren't.

Reply:
Oh true, hadn't thought of that. I'm not sure what would be the best way to handle this situation, maybe a static lock to prevent two `swaps` from happening at the same time?

I think they aren't recursive because I haven't found an initialization of `m_varLock`, so it's probably using the default constructor which is non-recursive.

Reply:
Not a best solution, but you can lock one with lower memory address and then higher. And do nothing if you're swapping two identical objects.

Reply:
Or you lock one and unlock it if you can't lock another one immediately.

Reply:
I found that there are [`std::lock()`](https://en.cppreference.com/w/cpp/thread/lock) and [`std::scoped_lock`](https://en.cppreference.com/w/cpp/thread/scoped_lock)(available in C++17), but `QReadWriteLock` isn't compatible with them. The algorithm looks like:
- Lock 1st mutex
- Try locking 2, 3, ..., Nth mutex without blocking
  - Success: good
  - Failure: Unlock 1st, shuffle the locking order, and retry

You can try implementing this.

Reply:
Always learning something new about locks! But yeah, I think you should be able to reimplement the algorithm used by `std::lock`, need to use `tryLockForWrite` instead then. I'm just confused as to what to do if it fails, since `swap` is `noexcept`. Just keep trying constantly? I wonder if that could in a very specific situation cause a deadlock anyways by the failures being synchronized:

- Thread A tries swap(x,y). Successfully locks X, fails to lock Y. Simultaneously thread B tries swap(y,x). Successfully locks Y, fails to lock X.
- They both unlock their locked objects and try again, but because they will try again in sync they end up in the same situation again.

I know introducing another lock might be unwanted, but maybe having a static lock called `SampleBuffer::s_swapLock`, and whichever thread calls `swap` needs to lock it, so we can be sure two swaps won't happen simultaneously.

Reply:
It's possible to successfully lock both of them thanks to some pseudo-randomness in OS scheduling.

Reply:
I was a concerbed about noexcept as well,and also came to the conclusion that a loop is the only way to solve this, but reintroduces the risk. It feels overkill, but I'm thinking of implementing exponential backoff to handle that.

Reply:
Interesting, didn't know about that! Is it done exclusively to avoid that kind of situation?

So @Spekular, I think the changes would be:
- Making `m_varLock` recursive by explicitly calling the QReadWriteLock constructor.
- Having this logic instead there:
```
bool successfulLock = false;
while(!succesfulLock)
{
    if (first.m_varLock.tryLockForWrite())
    {
        if (second.m_varLock.tryLockForWrite())
        {
            successfulLock = true;
        }
        else
        {
            first.m_varLock.unlock();
        }
    }
}
```

That's my first thought on how to approach, maybe there's a better way?

Reply:
> Thread A calls `swap(x, y)` while Thread B calls `swap (y, x)`

How about we just don't do this? There's no requirement for this to be possible; most other types don't support it.

Reply:
It's not like I was planning on doing it, I just can't say for sure that it will never happen.

Reply:
Other types can't say for sure that it won't happen either. C++ isn't idiot-proof - if you write silly code, you get silly results. `std::unique_ptr` is supposed to safely manage pointers, but that doesn't mean it stops you doing something like `delete p.get();` and getting hit in the face later with a double-free. I don't think it's unreasonable to say "swapping two of these objects twice simultaneously will bring pain and misery (just like most objects in C++)," and let people experience pain and misery if they try it.

This situation isn't a concern with the code as it is at the moment anyway - `swap` is currently called from `SampleBuffer`'s assignment operator, where one of the parameters is always a variable local to that method. I also feel that it's a bit out of scope - this PR is to make `SampleBuffer` and `SampleTCO` follow the rule of three, but you're trying to make a function - one that wasn't part of the original plan - do something that such functions don't typically do anyway.

Reply:
It took me a while to understand your point @DomClark , because somehow I was thinking that `swap(a,c)` running simultaneously with `swap(b,c)` could be an issue (which is not an unlikely situation if we have `bufferA = bufferC;` and `bufferB = bufferC;`). But in this case there's no deadlock, because even if the lock from `C` is disputed by both swaps, the other one is not, so we don't end up in a deadlock. The only situation where it would be an issue would be if we had `bufferA = bufferB;` and `bufferB = bufferA;` which indeed looks like bad code already if they are running concurrently.

It could be some sort of Poka Yoke to have the `tryLockForWrite` logic there though, but I see that it shouldn't be necessary as you said. You think we should keep the lock calls as they are and just make the lock recursive?

Reply:
If swap is thread unsafe by convention then let's just leave this code as is. I just defaulted to trying to be defensive, because preventing hold and wait here would be simpler than making a resource allocation graph for all of LMMS.

Reply:
The address solution is pretty simple though, so unless it's broken or has negative consequences I'll just use that.

Reply:
Implemented in recent commit. Feel free to unresolve if you think I should revert this and just do them in argument order, @DomClark.